### PR TITLE
feat: generate tailcall config from rest endpoint.

### PR DESCRIPTION
**Pull request by laststylebender14**
**Summary:**  
_Briefly describe the changes made in this PR._

**Issue Reference(s):**  
Fixes #... _(Replace "..." with the issue number)_

**Build & Testing:**

- [ ] I ran `cargo test` successfully.
- [ ] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [ ] I have added relevant unit & integration tests.
- [ ] I have updated the [documentation] accordingly.
- [ ] I have performed a self-review of my code.
- [ ] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

### CODE REVIEW

File: src/core/generator/from_json.rs [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+                }
+            }
+        "#;
+
+        let resp = serde_json::from_str(resp).unwrap();
+        let cgf_module = ConfigModule::from(from_json("https://example.com", &resp).config);
+        insta::assert_snapshot!(cgf_module.to_sdl());
+    }
+
+    #[test]
```
Comment (ranjitmahadik) [inline src/core/generator/from_json.rs:286]: add snap having query params. to verify mustache templates are correctly generated.

File: src/cli/tc.rs [hunk 2]
```diff
@@ -82,14 +83,28 @@ pub async fn run() -> Result<()> {
...
+                    // TODO: once the input format is discussed, move this to Generator.
+                    for url in paths {
+                        let parsed_url = Url::parse(&url).expect("failed to parse the url");
+                        let request = reqwest::Request::new(Method::GET, parsed_url);
+                        let resp = runtime.http.execute(request).await?.to_json()?;
+                        let config = from_json(&url, &resp.body);
+                        Fmt::display(config.to_sdl());
+                    }
+                }
+            }
```
Comment (tusharmath) [inline src/cli/tc.rs:107]: Move this code to generator.

File: src/cli/command.rs [hunk 3]
```diff
@@ -62,7 +62,7 @@ pub enum Command {
     Gen {
         /// Path of the source files separated by spaces if more than one
         #[arg(required = true)]
-        file_paths: Vec<String>,
+        paths: Vec<String>,
```
Comment (tusharmath) [inline src/cli/command.rs:65]: update docs on the website repo

File: src/core/generator/from_json.rs [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+                    "name": "Testing",
+                    "container": {
+                        "name": "Testing",
+                        "container": {
+                            "age": 16
+                        }
+                    }
+                }
+            }
+        "#;
```
Comment (tusharmath) [inline src/core/generator/from_json.rs:278]: - Writing JSON in Rust is a little tough, consider writing the inputs in `jsonc` or `YML` files for eg:

    ```json
    {
    "request": {
        "url": "http://example.com/api/endpoint"
    },
    "response": {
        "status": 200,
        "body": {}
    }
    }
    ```
- Use `datatest` to generate all the tests.

File: src/cli/tc.rs (hunk 2 above)
Comment (ranjitmahadik) [reply src/cli/tc.rs:100]: done.

File: src/cli/command.rs (hunk 3 above)
Comment (ranjitmahadik) [reply src/cli/command.rs:65]: will raise pr for doc change once this pr get's approved.

File: src/core/generator/from_json.rs (hunk 4 above)
Comment (ranjitmahadik) [reply src/core/generator/from_json.rs]: done, have moved hardcoded json files to data tests.

File: tests/execution_spec.rs [hunk 5]
```diff
@@ -9,4 +10,11 @@ fn run_execution_spec(path: &Path) -> datatest_stable::Result<()> {
...
 
-datatest_stable::harness!(run_execution_spec, "tests/execution", r"^.*\.md$");
+datatest_stable::harness!(
+    run_execution_spec,
+    "tests/execution",
+    r"^.*\.md$",
+    run_json_to_config_spec,
+    "tailcall-fixtures/fixtures/json",
+    r"^.*\.json"
+);
```
Comment (tusharmath) [inline tests/execution_spec.rs:20]: use datatests but don't write them here. Move it to the generator module.

File: src/core/generator/generator.rs [hunk 6]
```diff
@@ -26,43 +32,86 @@ fn resolve_file_descriptor_set(descriptor_set: FileDescriptorSet) -> Result<File
...
-        config.links = links;
-        Ok(
-            ConfigModule::from(config).resolve_ambiguous_types(|v| Resolution {
-                input: format!("{}Input", v),
-                output: v.to_owned(),
-            }),
-        )
+                Ok(configs)
+            }
+        }
```
Comment (tusharmath) [inline src/core/generator/generator.rs:114]: Need to be consistent with GRPC. For now let's generate only one merged config.

File: tailcall-fixtures/fixtures/json/list.json [hunk 7]
```diff
@@ -0,0 +1,9 @@
+{
+  "url": "https://example.com/users",
+  "body": [
+    {"name": "test", "age": 12},
+    {"name": "test-1", "age": 19},
+    {"name": "test-3", "age": 21, "adult": true},
+    {"name": "test-3", "age": 21}
+  ]
+}
```
Comment (tusharmath) [inline tailcall-fixtures/fixtures/json/list.json:9]: Move the test inputs to generator.

File: src/core/generator/from_json.rs [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+                key: k.to_string(),
+                data_type: detect_gql_data_type(&v),
+                is_list: v.contains(","),
+            })
+            .collect();
+        Self { queries: query_list }
+    }
+}
+
+struct ConfigGenerator {
```
Comment (tusharmath) [inline src/core/generator/from_json.rs:37]:
```rust
trait ConfigTransformer {
    fn transform(&self, config: Config) -> Config
}

// json::upstream.rs
struct InsertUpstream
impl ConfigTransformer for InsertUpstream {}

// json::server.rs
struct InsertServer {  }
impl ConfigTransformer for InsertServer {}

// json::types.rs
struct InsertTypes
impl ConfigTransformer for InsertTypes {}

// json::fields.rs
struct InsertFields
impl ConfigTransformer for InsertFields {}

```

Can structure this module a little better.

File: src/core/generator/json/snapshots/tailcall__core__generator__json__query_generator__test__list_json_query_generator.snap [hunk 9]
```diff
@@ -0,0 +1,7 @@
+---
+source: src/core/generator/json/query_generator.rs
+expression: config.to_sdl()
+---
+type Query {
+  f1: [T1] @http(path: "/path")
```
Comment (tusharmath) [inline src/core/generator/json/snapshots/tailcall__core__generator__json__query_generator__test__list_json_query_generator.snap:6]: T1 is not defined in this configuraion?

**Review by tusharmath — APPROVED** [review]

**Review by tusharmath — APPROVED** [review]

**Pull request merged**